mlflow/langchain/__init__.py

```diff
@@ -1003,37 +1005,50 @@ def autolog(
             new model version of the registered model with this name.
             The registered model is created if it does not already exist.
         extra_tags: A dictionary of extra tags to set on each managed run created by autologging.
+        extra_log_classes: A list of langchain classes to log in addition to the default classes.
+            We do not guarantee classes specified in this list can be logged as a model, but tracing
+            will be supported. Note that all classes within the list must be subclasses of Runnable,
+            and we only patch `invoke`, `batch`, and `stream` methods for tracing.
     """
-
     with contextlib.suppress(ImportError):
+        import langchain
+        import langchain_community
         from langchain.agents.agent import AgentExecutor
         from langchain.chains.base import Chain
         from langchain.schema import BaseRetriever
+        from langchain.schema.runnable import Runnable
 
-        classes = lc_runnables_types() + (AgentExecutor, Chain)
-        for cls in classes:
-            # If runnable also contains loader_fn and persist_dir, warn
-            # BaseRetrievalQA, BaseRetriever, ...
-            safe_patch(
-                FLAVOR_NAME,
-                cls,
-                "invoke",
-                functools.partial(patched_inference, "invoke"),
-            )
-
-            safe_patch(
-                FLAVOR_NAME,
-                cls,
-                "batch",
-                functools.partial(patched_inference, "batch"),
-            )
-
-            safe_patch(
-                FLAVOR_NAME,
-                cls,
-                "stream",
-                functools.partial(patched_inference, "stream"),
-            )
+        def _patch_runnable_cls(cls):
+            for func_name in ["invoke", "batch", "stream"]:
+                if hasattr(cls, func_name):
+                    safe_patch(
+                        FLAVOR_NAME, cls, func_name, functools.partial(patched_inference, func_name)
+                    )
+                    logger.debug(f"patched {cls.__name__}.{func_name}")
+
+        def _inspect_module_and_patch_cls(module):
+            for _, obj in inspect.getmembers(module):
+                if inspect.ismodule(obj) and (obj.__name__.startswith("langchain")):
+                    _inspect_module_and_patch_cls(obj)
+                elif inspect.isclass(obj) and issubclass(obj, Runnable):
+                    _patch_runnable_cls(obj)
```

Comment:
Any chance that this leads to duplicate patching?

```python
class Runnable:
    def invoke(...)


class MyRunnable(Runnable):
    def invoke(...)


class YourRunnable(MyRunnable):
    def invoke(...)
```

Reply:
Runnable's `invoke` method is an abstract method, seems like in langchain most of the classes just implement the invoke method or some other internal method instead of calling `super().invoke`. Is there anyway to avoid such duplication if it exists? I'll add patched classes into a set in case we patch the same class twice though!

Reply:
Can we check if this really occurs by creating a custom class that inherits the existing class?

Reply:
Duplicate patching is acceptable as long as `YourRunnable.invoke` triggers autologging in the example above.

Reply:
sklearn autologging has a mechanism to prevent this using `_TrainingSession`.